### PR TITLE
Adding external dependency on libmemcached to gstpufs

### PR DESCRIPTION
**Pull request by marioecd**
gstpufs will run a memcached file block cache that has an external dependency on libmemcached. This PR adds memcached_dao_interface and makes sure that it can build with the external dependency on libmemcached.

### CODE REVIEW

**Comment by marioecd** [comment]
cla issue solved. The seemingly unrelated (to me) bazel windows check still failed.

**Comment by marioecd** [comment]
Only Bazel Windows and Test 3.7 MacOS continue to fail. The errors look to me like they are unrelated to my PR.

**Comment by yongtang** [comment]
@marioecd The windows failure is unrelated, though it could be fixed with:
```diff
--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -319,7 +319,7 @@ jobs:
         shell: cmd
         run: |
           @echo on
-          choco install bazel --version=%BAZEL_VERSION%
+          curl -sSL -o bazel.exe https://github.com/bazelbuild/bazel/releases/download/%BAZEL_VERSION%/bazel-%BAZEL_VERSION%-windows-x86_64.exe
           bazel version
           cp /c/hostedtoolcache/windows/Python/%PYTHON_VERSION%/x64/python /c/hostedtoolcache/windows/Python/%PYTHON_VERSION%/x64/python3
           python3 --version
```

File: tensorflow_io/core/BUILD [hunk 1]
```diff
@@ -572,6 +572,7 @@ cc_binary(
         "@local_config_tf//:tf_header_lib",
     ] + select({
         "@bazel_tools//src/conditions:windows": [],
+        "@bazel_tools//src/conditions:darwin": [],
```
Comment (yongtang) [inline tensorflow_io/core/BUILD:575]: The macOS failure is actually caused by this line. Let me think how to address it while skipping the libmemcached for macOS.

**Comment by yongtang** [comment]
@marioecd The following with fix the macOS issue (basically, copy content of `default` to `darwin` selection, minus the `tpu`):
```diff
diff --git a/tensorflow_io/core/BUILD b/tensorflow_io/core/BUILD
index 1c0684b..ae000d9 100644
--- a/tensorflow_io/core/BUILD
+++ b/tensorflow_io/core/BUILD
@@ -572,7 +572,13 @@ cc_binary(
         "@local_config_tf//:tf_header_lib",
     ] + select({
         "@bazel_tools//src/conditions:windows": [],
-        "@bazel_tools//src/conditions:darwin": [],
+        "@bazel_tools//src/conditions:darwin": [
+            "//tensorflow_io/core:core_ops",
+            "//tensorflow_io/core:genome_ops",
+            "//tensorflow_io/core:optimization",
+            "//tensorflow_io/core/oss:oss_ops",
+            "//tensorflow_io/gcs:gcs_config_ops",
+        ],
         "//conditions:default": [
             "//tensorflow_io/core:core_ops",
             "//tensorflow_io/core:genome_ops",
```

File: WORKSPACE [hunk 2]
```diff
@@ -10,6 +10,10 @@ load("//third_party/toolchains/gpu:cuda_configure.bzl", "cuda_configure")
 
 cuda_configure(name = "local_config_cuda")
 
+load("//tensorflow_io/core/kernels/gstpufs:workspace.bzl", "gstpufs_repositories")
+
+gstpufs_repositories()
```
Comment (yongtang) [inline WORKSPACE:15]: Those two lines are not needed. The `workspace.bzl` style in TF core are mostly comes from Bazel's pre-1.0 days when many features are not available in Bazel. So TF was forced to patch around with different `.bzl` files.

Now bazel is in 3.0+, just copy the content from `workspace.bzl` to here, remove workspace.bzl, and remove the above two lines will be fine:
```
    http_archive(
         name = "libmemcached",
         urls = [
             "https://storage.googleapis.com/mirror.tensorflow.org/launchpad.net/libmemcached/1.0/1.0.18/+download/libmemcached-1.0.18.tar.gz",
             "https://launchpad.net/libmemcached/1.0/1.0.18/+download/libmemcached-1.0.18.tar.gz",
         ],
         sha256 = "e22c0bb032fde08f53de9ffbc5a128233041d9f33b5de022c0978a2149885f82",
         strip_prefix = "libmemcached-1.0.18",
         build_file = "//third_party:libmemcached.BUILD",
     )
```

**Pull request merged**